FC-0001: Remove Persistent Course Grades Flags

UvgenGen opened the pull request:
## Description
Persistence grade removing is accepted and in the removing stage
https://github.com/openedx/public-engineering/issues/29

Remove the PersistentGradesEnabledFlag, CoursePersistentGradesFlag, and the related waffle flag, ASSUME_ZERO_IF_ABSENT. We are persisting the grades and assuming zero if absent.

After applying this change, courses for which the flag was turned off should not break. But to display the correct grade, you will need to run the command to recalculate grades:
` ./manage.py lms compute_grades --all_courses`

PR with removing the model with a migration: https://github.com/openedx/edx-platform/pull/30977

UvgenGen committed:
feat: revert Fix certificate generation without persistent grades

openedx-webhooks commented:
Thanks for the pull request, @UvgenGen!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.




<!-- comment:external_pr -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

UvgenGen committed:
feat: Remove Persistent Course Grades Flags

UvgenGen committed:
test: update tests

schenedx commented on common/djangoapps/util/tests/test_db.py:
```diff
@@ -120,6 +121,7 @@ class MigrationTests(TestCase):
     Tests for migrations.
     """
 
+    @skip
```
We should have this changed back when you introduce the migration file with the next PR, right?

UvgenGen replied:
Yes, we will changed back this with the next PR.

schenedx commented on lms/djangoapps/course_home_api/progress/tests/test_views.py:
```diff
@@ -193,48 +192,54 @@ def test_learner_has_access(self):
 
-    @patch.dict(settings.FEATURES, {'ASSUME_ZERO_GRADE_IF_ABSENT_FOR_ALL_TESTS': False})
     def test_override_is_visible(self):
-        with persistent_grades_feature_flags(global_flag=True):
```
Why remove these items in tests? We should leave these in, but outside of the `with` statement right? We want to make the persistentGrade to be true all the time after this change, right?

UvgenGen replied:
it's right, persistGrade value is always true now, so persistent_grades_feature_flags was removed in this test.
Also completely removed the [persist_grades_feature_flags](https://github.com/openedx/edx-platform/pull/30978/files#diff-9e7e3de020c6b1214b2023343184790e428ef034deff90acd07f83adc0f54063L15) function.

schenedx replied:
yes. I agree that the `with` statement should be removed.
However, why remove line 199 -202? Shouldn't we leave that in for this test? If we leave those in, we can further validate the subsection grade persistent, right?

schenedx reviewed (changes requested):
I see many unit test changes where, the code should not be removed because the code condition was with flags set to True. We want our grading logic to persist grades into db and assume_zero by default, right?

schenedx commented on lms/djangoapps/grades/course_grade.py:
```diff
@@ -294,14 +293,7 @@ def attempted(self):  # lint-amnesty, pylint: disable=invalid-overridden-method
-                if subsection_grade.all_total.first_attempted:
-                    return True
-        return False
+        return True
```
I don't like this change at all. I understand logically, this is correct. However, this makes the meaning of the `attempted` property useless. Can we go to all places where this particular property is referenced and update the logic to not use the property? That way, we can remove this property all together.

UvgenGen replied:
I thought about it. But this function just override this [method from CourseGradeBase](https://github.com/openedx/edx-platform/pull/30978/files#diff-f0d17d91bec07c249a69bfeb1b4be40078e8c433136af33644f429eeddcab9b3R56). CourseGradeBase is parent for ZeroCourseGrade and CourseGrade, so this method return `False` for ZeroCourseGrade.

schenedx replied:
Ah, I see. This makes sense then.

schenedx commented on lms/djangoapps/grades/tests/test_course_grade_factory.py:
```diff
@@ -54,57 +46,6 @@ def test_course_grade_no_access(self):
         assert grade.percent == 0
 
-    @patch.dict(settings.FEATURES, {'PERSISTENT_GRADES_ENABLED_FOR_ALL_TESTS': False})
-    @ddt.data(
```
These tests are removed because the value `PERSISTENT_GRADES_ENABLED_FOR_ALL_TESTS` should be `True` all the time now, right? Are there tests within this file covers the case where `PERSISTENT_GRADES_ENABLED_FOR_ALL_TESTS` is `True`? I am mindful of test coverage percentage.

UvgenGen replied:
Yes, these tests are removed because the value `PERSISTENT_GRADES_ENABLED_FOR_ALL_TESTS` should be `True` all the time now.
`PERSISTENT_GRADES_ENABLED_FOR_ALL_TESTS` feature removed from platform at all. This feature used in [PersistentGradesEnabledFlag feature_enabled](https://github.com/openedx/edx-platform/pull/30978/files#diff-61855a03b2acfd3689924fa13918008b4dc22dee0089f75256568bb8fe1d2ebcL43) function. 
I believe [test_read_and_update](https://github.com/openedx/edx-platform/pull/30978/files#diff-d009728fd48d9c7ca695f14cb0e3f71c712aab9ba68d8b7390a36bb40d210d6bL108) covers this case.

schenedx commented:
@UvgenGen Left some more comments after my second round of review. I am wondering if you have done any manual testing on your local machine for this change?
Please let me know after you have addressed all my questions/concerns.

UvgenGen commented:
@schenedx I've addressed your comments.
About manual testing on my local machine, I only checked flow with creation the course -> setting grade for the course -> passing the course -> checking user grade. It's work as expected.

schenedx reviewed (approved):
OK, with the recent round of changes. I loaded this branch onto my local machine. I tested it briefly myself and it does look good. 
Merging 1 hour later

schenedx commented:
@UvgenGen For the comment, I don't this that works for edx.org. We have 10k courses that might be in this situation.
Can you clarify what you mean by the course should not break? Also, what is the best way to have course grade or subsection grade filled in gradually and on-demand? Running that management command on edx.org production causes severe performance penalty to most of edx.org systems.

> After applying this change, courses for which the flag was turned off should not break. But to display the correct grade, you will need to run the command to recalculate grades:
./manage.py lms compute_grades --all_courses

schenedx reviewed (changes requested):
Toggle the approval back to request changes because of the backfill method described in the PR description.

schenedx reviewed (approved):
With further discussion with my team members, I believe this PR should introduce no significant harm to edx.org environments. Green light this PR again.

openedx-webhooks commented:
@UvgenGen 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/edx-platform&entry.867055334=https://github.com/openedx/edx-platform/pull/30978&entry.1484655318=https://github.com/UvgenGen&entry.752974735=2022-09-12+14:37&entry.1917517419=2022-09-14+18:00&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->
